d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,200 @@
+        return;
+      }
+
+      cluster._matched = 0;
+
+      if (cluster._oldLb == null || cluster._newLb == null) {
+        LOG.debug("Added new URI properties for {} for {} LB.", clusterName, fromNewLb ? "New" : "Old");
+
+        cluster._uris = (cluster._oldLb == null) ? cluster._newLb.Uris().size() : cluster._oldLb.Uris().size();
+        _totalUris += cluster._uris;
+
+        updateJmxMetrics(clusterName, cluster);
+        return;
+      }
+
+      cluster._uris = cluster._oldLb.Uris().size();
+      Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+      for (URI uri : cluster._oldLb.Uris()) {
+        if (!newLbUris.remove(uri)) {
+          continue;
+        }
+
+        if (compareURI(uri, cluster._oldLb, cluster._newLb)) {
+          cluster._matched++;
+        }
+      }
+      // add the remaining unmatched URIs in newLbUris to the uri count
+      cluster._uris += newLbUris.size();
+
+      if (cluster._matched != cluster._uris) {
+        RATE_LIMITED_LOGGER.info("Mismatched cluster properties for {} (match score: {}, total uris: {}):"
+                + "\nOld LB: {}\nNew LB: {}",
+            clusterName, cluster._matched / cluster._uris, cluster._uris, cluster._oldLb, cluster._newLb);
+      }
+
+      _totalUris += cluster._uris;
+      _matchedUris += cluster._matched;
+
+      updateJmxMetrics(clusterName, cluster);
+    }
+  }
+
+  private void updateJmxMetrics(String clusterName, ClusterMatchRecord cluster) {
+    // set a clone of cluster match record to jmx to avoid jmx reading the record in the middle of an update
+    _dualReadLoadBalancerJmx.setClusterMatchRecord(clusterName, cluster == null ? null : cluster.clone());
+    _dualReadLoadBalancerJmx.setUriPropertiesSimilarity((double) _matchedUris / (double) _totalUris);
+  }
+
+  private static boolean compareURI(URI uri, UriProperties oldLb, UriProperties newLb) {
+    String clusterName = oldLb.getClusterName();
+    return compareMaps("partition desc", clusterName, uri, UriProperties::getPartitionDesc, oldLb, newLb) &&
+        compareMaps("specific properties", clusterName, uri, UriProperties::getUriSpecificProperties, oldLb, newLb);
+  }
+
+  private static <K, V> boolean compareMaps(
+      String type, String cluster, URI uri, Function<UriProperties, Map<URI, Map<K, V>>> extractor,
+      UriProperties oldLb, UriProperties newLb
+  ) {
+    Map<K, V> oldData = extractor.apply(oldLb).get(uri);
+    Map<K, V> newData = extractor.apply(newLb).get(uri);
+    if (Objects.equals(oldData, newData)) {
+      return true;
+    }
+
+    LOG.debug("URI {} for {}/{} mismatched between old and new LB.\nOld LB: {}\nNew LB: {}",
+        type, cluster, uri, oldData, newData);
+    return false;
+  }
+
+  @VisibleForTesting
+  int getTotalUris() {
+    return _totalUris;
+  }
+
+  @VisibleForTesting
+  int getMatchedUris() {
+    return _matchedUris;
+  }
+
+  public static class ClusterMatchRecord implements Cloneable {
```

Comment:
I wouldn't even bother with `Cloneable` here. I never really liked it and we don't know what it actually does. For example, does it do a deep clone? If so, it's making full copies of the `UriProperties` for no reason. IMO just have a constructor:
```java
private ClusterMatchRecord(UriProperties oldLb, UriProperties newLb, int uris, int matched) {...}
```
You can have a no-arg constructor too, for convenience:
```
private ClusterMatchRecord() {}
```
This one doesn't need to invoke `this(...)` since all the fields always start out as the 0-value.

Reply:
sure (.clone() doc says it does a shallow copy btw, but yeah, not using it now)